pytorch_lightning/callbacks/pt_callbacks.py

```diff
@@ -225,29 +246,46 @@ def save_model(self, filepath, overwrite):
 
     def on_epoch_end(self, epoch, logs=None):
         logs = logs or {}
-        self.epochs_since_last_save += 1
-        if self.epochs_since_last_save >= self.period:
-            self.epochs_since_last_save = 0
+        self.epochs_since_last_check += 1
+        if self.epochs_since_last_check >= self.period:
+            self.epochs_since_last_check = 0
             filepath = '{}/{}_ckpt_epoch_{}.ckpt'.format(self.filepath, self.prefix, epoch + 1)
-            if self.save_best_only:
+            if self.save_top_k:
                 current = logs.get(self.monitor)
                 if current is None:
                     print('Can save best model only with %s available,'
                           ' skipping.' % (self.monitor), RuntimeWarning)
                 else:
-                    if self.monitor_op(current, self.best):
+                    if ((len(self.bestk.keys()) < self.save_top_k) or
+                            (self.monitor_op(current, self.bestk[self.kth]))):
+                        if len(self.bestk.keys()) == self.save_top_k:
+                            # need to pop the kth
+                            delpath = '{}/{}_ckpt_epoch_{}.ckpt'.format(
+                                self.filepath, self.prefix, self.kth + 1)
+                            self.bestk.pop(self.kth)
+                            self.del_model(delpath)
+                        self.bestk[epoch] = current
+                        if len(self.bestk.keys()) == self.save_top_k:
+                            # monitor dict has reached k elements
+                            if self.mode == 'min':
+                                self.kth = max(self.bestk, key=self.bestk.get)
+                            else:
+                                self.kth = min(self.bestk, key=self.bestk.get)
+                        if self.mode == 'min':
+                            self.best = min(self.bestk.values())
+                        else:
+                            self.best = max(self.bestk.values())
                         if self.verbose > 0:
-                            print('\nEpoch %05d: %s improved from %0.5f to %0.5f,'
-                                  ' saving model to %s'
-                                  % (epoch + 1, self.monitor, self.best,
-                                     current, filepath))
-                        self.best = current
+                            print('\nEpoch %05d: %s reached %s (best %s),'
```

Comment:
consider using `logging` instead of `print`
```python
logging.info('Epoch %05d: %s reached %s (best %s), saving model to %s as top %d',
             epoch + 1, self.monitor, current, self.best, filepath, self.save_top_k)
```

Reply:
this is a change beyond this pr, the whole file is using print, I think you should create a new issue for this